### ISSUE TITLE
Memoized values

### ISSUE THREAD

**Issue by raniejade** [JetBrains/spek#102]
Basically they are user defined values whose lifetime is bound to either to a `group` or `test`.  Meaning each group/test can have their own instance of that value. Subject (#87) is implemented this way and we can reuse some stuff. 

The DSL proposed requires Kotlin 1.1 as it will use local delegated properties.

``` kotlin
interface Dsl {
    ...
    fun <T> memoized(mode: CachingMode = CachingMode.TEST, factory: () -> T): Memoized<T>
}
```

``` kotlin
class SomeSpek: Spek({
    val foo by memoized {
        // create foo
    }

    it("this is a test") {
        // first access will invoke the factory
        foo.doSomething()
    }
})
```

It's also possible to extend this for example (to resolve #95):

``` kotlin
fun <T: Closeable> Dsl.resource(mode: CachingMode = CachingMode.TEST: factory() -> T): Memoized<T> {
    val memoized = memoized(mode, factory)
    val resource by memoized
    when(mode) {
        CachingMode.TEST -> afterEach { resource.close() }
        CachingMode.GROUP -> after { resource.close() }
    }
    return memoized
}
```

**Comment by craigjbass** [comment]
Trying to get this working with Kotlin 1.1 EAP

Then I noticed a discrepancy between the release version on GitHub (1.0.0) and the maven repo version (1.0.9)

And like, I'm well confused

**Comment by craigjbass** [comment]
I was playing around a bit and got something working in Spek 1.0.9

```kotlin
package some.package

import org.amshove.kluent.`should equal`
import org.jetbrains.spek.api.DescribeBody
import org.jetbrains.spek.api.Spek
import kotlin.reflect.KProperty

class SpekSpek : Spek({
    describe("Speks") {
        val valueStorage by memoized { ValueStorage() }
        val tester by memoized { Tester(0, valueStorage) }

        beforeEach {
            valueStorage.call()
        }
        context( "not called" ) {
            it("Spek") {
                tester.a.`should equal`(0)
                valueStorage.b.`should equal`(1)
            }
        }

        context( "called once" ) {
            beforeEach {
                tester.call()
            }
            it("Spek") {
                tester.a.`should equal`(1)
                valueStorage.b.`should equal`(1)
            }
        }

        context( "called twice" ) {
            beforeEach {
                tester.call()
                tester.call()
            }
            it("Spek") {
                tester.a.`should equal`(2)
                valueStorage.b.`should equal`(1)
            }
            it("Spek") {
                tester.a.`should equal`(2)
                valueStorage.b.`should equal`(1)
            }
        }

        context( "called twice" ) {
            val valueStorage by memoized { ValueStorage() }
            beforeEach {
                tester.call()
                tester.call()
                valueStorage.call()
            }
            it("Spek") {
                tester.a.`should equal`(2)
            }
            it("Spek") {
                tester.a.`should equal`(2)
                valueStorage.b.`should equal`(1)
            }
        }
    }
})

class Tester(public var a: Int = 0, public var bStorage: ValueStorage) {
    fun call() {
        a++
    }
}

class ValueStorage(public var b: Int = 0) {
    fun call() {
        b++
    }
}

class Delegate<T>(val factory: () -> T) {
    private var value: T? = null
    operator fun getValue(thisRef: Nothing?, property: KProperty<*>): T {
        value = value ?: factory()
        return value !!
    }
    fun clear() {
        value = null
    }
    operator fun setValue(thisRef: Any?, property: KProperty<*>, value: T) {}
}

fun <T> DescribeBody.memoized(factory:() -> T): Delegate<T> {
    val delegate = Delegate(factory)
    afterEach {
        delegate.clear()
    }
    return delegate
}
```
![image](https://cloud.githubusercontent.com/assets/1889973/18420056/ae75abbe-7861-11e6-90d5-6ecc327c84af.png)

![image](https://cloud.githubusercontent.com/assets/1889973/18420053/9bdfcdcc-7861-11e6-9aad-5edd3392ccb0.png)

**Comment by craigjbass** [comment]
Although it is difficult to recreate the behaviour of let() from RSpec....

```kotlin
package uk.co.devls.spek

import org.amshove.kluent.`should equal`
import org.jetbrains.spek.api.Spek


class SpekSpek : Spek({
    describe("spek memoization in kotlin 1.1 EAP") {
        val nestedObject by memoized { ValueStorage() }
        val testSubject by memoized { Tester(0, nestedObject) }

        context("test subject is not called") {
            it("test subject value should be 0") { testSubject.a.`should equal`(0) }
            it("nested object should contain value 0") { testSubject.b().`should equal`(0) }
            context("and nested object is called") {
                beforeEach { nestedObject.increment() }
                it("nested object should contain value 1") { testSubject.b().`should equal`(1) }
                it("subject should contain value 0") { testSubject.a.`should equal`(0) }
            }
        }

        context("test subject is called once") {
            beforeEach { testSubject.increment() }
            it("test subject should contain value 1") { testSubject.a.`should equal`(1) }
            it("nested object should contain value 1") { testSubject.b().`should equal`(0) }
        }

        context("test subject is called twice") {
            beforeEach {
                testSubject.increment()
                testSubject.increment()
            }
            it("test subject should contain value 2") { testSubject.a.`should equal`(2) }
            it("nested object should contain value 0") { testSubject.b().`should equal`(0) }
        }

        context("when nested object is shadowed") {
            val nestedObject by memoized { ValueStorage() }
            beforeEach {
                testSubject.increment()
                testSubject.increment()
                nestedObject.increment()
            }
            it("should contain 1") {
                testSubject.b().`should equal`(1)
            }
            context("and incremented again") {
                beforeEach {
                    testSubject.increment()
                    testSubject.increment()
                    nestedObject.increment()
                }
                it("should contain 2") { testSubject.b().`should equal`(2) }
            }
        }
    }
})

class Tester(public var a: Int = 0, public var bStorage: ValueStorage) {
    fun increment() {
        a++
    }

    fun b(): Int {
        return bStorage.b
    }
}

class ValueStorage(public var b: Int = 0) {
    fun increment() {
        b++
    }
}
```

![image](https://cloud.githubusercontent.com/assets/1889973/18420207/f08e41f2-7864-11e6-8298-3a9eb4fdec13.png)

**Comment by raniejade** [comment]
@craigjbass Latest release is `1.0.89`, which rewrote most of the internals. Awesome work, if you checkout the latest master you will see a lot has changed.

**Comment by raniejade** [comment]
Resolved by b9967f3f0cdf1e31c14eeb83963be04b18d65220.
